subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/jdk7/NioFileMetadataAccessor.java

```diff
@@ -16,29 +16,41 @@
 package org.gradle.internal.nativeintegration.filesystem.jdk7;
 
 import org.gradle.internal.file.FileMetadataSnapshot;
+import org.gradle.internal.file.FileMetadataSnapshot.AccessType;
 import org.gradle.internal.file.impl.DefaultFileMetadataSnapshot;
 import org.gradle.internal.nativeintegration.filesystem.FileMetadataAccessor;
 
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
 import java.nio.file.attribute.BasicFileAttributes;
 
 @SuppressWarnings("Since15")
 public class NioFileMetadataAccessor implements FileMetadataAccessor {
     @Override
     public FileMetadataSnapshot stat(File f) {
         try {
-            BasicFileAttributes bfa = Files.readAttributes(f.toPath(), BasicFileAttributes.class);
+            Path path = f.toPath();
+            BasicFileAttributes bfa = java.nio.file.Files.readAttributes(path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
```

Comment:
It would be great to use proper variable names instead of `f` and `bfa`.

Reply:
We should wrap only this with a `try` block. Also, we should handle `NoSuchFileException` only; other `IOException`s should be rethrown as `UncheckedIOException`s.

Reply:
We talked about that yesterday: We treat all files which we can't stat as `missing`. That includes files which we don't have enough permissions to read and cyclic symlinks. So catching all `IOException`s here is exactly what we need to do to conform to the Javadoc of `FileMetadataAccessor.stat`: https://github.com/gradle/gradle/blob/6d18724153c738866ef4cc768585f4e0215bca8a/subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/FileMetadataAccessor.java#L24-L41

Both accesses need to be wrapped in a try block, because both can throw `IOException`s for different reasons.

Reply:
There are two things here. One is that we should only wrap the `readAttributes()` call not the rest. I feel strongly about doing this part.

The other is wether we should catch `IOException` or specific exceptions (i.e. `NoSuchFileException` and `AccessDeniedException`). I'm arguing that the latter is a better solution, as we are explicit about the problems we handle, and then we can recognize if we don't handle something. I feel less strongly about doing this.

Reply:
I did the first (tighten the scope of `try`).

We can't do the second, since that would break supporting cyclic symlink dependencies if I don't want to parse the error message.

Reply:
We could also catch `FileSystemLoopException` explicitly, making it obvious what actual cases we treat as missing.

Reply:
`readAttributes` does not throw `FileSystemLoopException` in the symlink loop case, only a generic `FileSystemException`. We would need to parse the message to find out that this was a symlink loop (`Too many levels of symbolic links`). Same goes for the `NativeException`. Which ones to catch and which ones to make stat fail.

The current behavior is that `stat` never throws an exception.